File: src/main/java/io/github/erp/service/FinancialDerivativeTypeCodeService.java
package io.github.erp.service;

/*-
 * Erp System - Mark VII No 3 (Gideon Series) Server ver 1.5.7
 * Copyright © 2021 - 2023 Edwin Njeru (mailnjeru@gmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import io.github.erp.service.dto.FinancialDerivativeTypeCodeDTO;
import java.util.Optional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service Interface for managing {@link io.github.erp.domain.FinancialDerivativeTypeCode}.
 */
public interface FinancialDerivativeTypeCodeService {
    /**
     * Save a financialDerivativeTypeCode.
     *
     * @param financialDerivativeTypeCodeDTO the entity to save.
     * @return the persisted entity.
     */
    FinancialDerivativeTypeCodeDTO save(FinancialDerivativeTypeCodeDTO financialDerivativeTypeCodeDTO);

    /**
     * Partially updates a financialDerivativeTypeCode.
     *
     * @param financialDerivativeTypeCodeDTO the entity to update partially.
     * @return the persisted entity.
     */
    Optional<FinancialDerivativeTypeCodeDTO> partialUpdate(FinancialDerivativeTypeCodeDTO financialDerivativeTypeCodeDTO);

    /**
     * Get all the financialDerivativeTypeCodes.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<FinancialDerivativeTypeCodeDTO> findAll(Pageable pageable);

    /**
     * Get the "id" financialDerivativeTypeCode.
     *
     * @param id the id of the entity.
     * @return the entity.
     */
    Optional<FinancialDerivativeTypeCodeDTO> findOne(Long id);

    /**
     * Delete the "id" financialDerivativeTypeCode.
     *
     * @param id the id of the entity.
     */
    void delete(Long id);

    /**
     * Search for the financialDerivativeTypeCode corresponding to the query.
     *
     * @param query the query of the search.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<FinancialDerivativeTypeCodeDTO> search(String query, Pageable pageable);
}

File: erp-client/src/main/webapp/app/entities/gdi/financial-derivative-type-code/service/financial-derivative-type-code.service.ts
///
/// Erp System - Mark X No 10 (Jehoiada Series) Client 1.7.8
/// Copyright © 2021 - 2024 Edwin Njeru (mailnjeru@gmail.com)
///
/// This program is free software: you can redistribute it and/or modify
/// it under the terms of the GNU General Public License as published by
/// the Free Software Foundation, either version 3 of the License, or
/// (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program. If not, see <http://www.gnu.org/licenses/>.
///

import { Injectable } from '@angular/core';
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Observable } from 'rxjs';

import { isPresent } from 'app/core/util/operators';
import { ApplicationConfigService } from 'app/core/config/application-config.service';
import { createRequestOption } from 'app/core/request/request-util';
import { SearchWithPagination } from 'app/core/request/request.model';
import { IFinancialDerivativeTypeCode, getFinancialDerivativeTypeCodeIdentifier } from '../financial-derivative-type-code.model';

export type EntityResponseType = HttpResponse<IFinancialDerivativeTypeCode>;
export type EntityArrayResponseType = HttpResponse<IFinancialDerivativeTypeCode[]>;

@Injectable({ providedIn: 'root' })
export class FinancialDerivativeTypeCodeService {
  protected resourceUrl = this.applicationConfigService.getEndpointFor('api/financial-derivative-type-codes');
  protected resourceSearchUrl = this.applicationConfigService.getEndpointFor('api/_search/financial-derivative-type-codes');

  constructor(protected http: HttpClient, protected applicationConfigService: ApplicationConfigService) {}

  create(financialDerivativeTypeCode: IFinancialDerivativeTypeCode): Observable<EntityResponseType> {
    return this.http.post<IFinancialDerivativeTypeCode>(this.resourceUrl, financialDerivativeTypeCode, { observe: 'response' });
  }

  update(financialDerivativeTypeCode: IFinancialDerivativeTypeCode): Observable<EntityResponseType> {
    return this.http.put<IFinancialDerivativeTypeCode>(
      `${this.resourceUrl}/${getFinancialDerivativeTypeCodeIdentifier(financialDerivativeTypeCode) as number}`,
      financialDerivativeTypeCode,
      { observe: 'response' }
    );
  }

  partialUpdate(financialDerivativeTypeCode: IFinancialDerivativeTypeCode): Observable<EntityResponseType> {
    return this.http.patch<IFinancialDerivativeTypeCode>(
      `${this.resourceUrl}/${getFinancialDerivativeTypeCodeIdentifier(financialDerivativeTypeCode) as number}`,
      financialDerivativeTypeCode,
      { observe: 'response' }
    );
  }

  find(id: number): Observable<EntityResponseType> {
    return this.http.get<IFinancialDerivativeTypeCode>(`${this.resourceUrl}/${id}`, { observe: 'response' });
  }

  query(req?: any): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http.get<IFinancialDerivativeTypeCode[]>(this.resourceUrl, { params: options, observe: 'response' });
  }

  delete(id: number): Observable<HttpResponse<{}>> {
    return this.http.delete(`${this.resourceUrl}/${id}`, { observe: 'response' });
  }

  search(req: SearchWithPagination): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http.get<IFinancialDerivativeTypeCode[]>(this.resourceSearchUrl, { params: options, observe: 'response' });
  }

  addFinancialDerivativeTypeCodeToCollectionIfMissing(
    financialDerivativeTypeCodeCollection: IFinancialDerivativeTypeCode[],
    ...financialDerivativeTypeCodesToCheck: (IFinancialDerivativeTypeCode | null | undefined)[]
  ): IFinancialDerivativeTypeCode[] {
    const financialDerivativeTypeCodes: IFinancialDerivativeTypeCode[] = financialDerivativeTypeCodesToCheck.filter(isPresent);
    if (financialDerivativeTypeCodes.length > 0) {
      const financialDerivativeTypeCodeCollectionIdentifiers = financialDerivativeTypeCodeCollection.map(
        financialDerivativeTypeCodeItem => getFinancialDerivativeTypeCodeIdentifier(financialDerivativeTypeCodeItem)!
      );
      const financialDerivativeTypeCodesToAdd = financialDerivativeTypeCodes.filter(financialDerivativeTypeCodeItem => {
        const financialDerivativeTypeCodeIdentifier = getFinancialDerivativeTypeCodeIdentifier(financialDerivativeTypeCodeItem);
        if (
          financialDerivativeTypeCodeIdentifier == null ||
          financialDerivativeTypeCodeCollectionIdentifiers.includes(financialDerivativeTypeCodeIdentifier)
        ) {
          return false;
        }
        financialDerivativeTypeCodeCollectionIdentifiers.push(financialDerivativeTypeCodeIdentifier);
        return true;
      });
      return [...financialDerivativeTypeCodesToAdd, ...financialDerivativeTypeCodeCollection];
    }
    return financialDerivativeTypeCodeCollection;
  }
}
